From: Jiri Slaby <jslaby@suse.cz>
Date: Tue, 19 May 2015 16:02:07 +0200
Subject: kgr: use kgr_in_progress for all threads
Patch-mainline: submitted for review
References: fate#313296 bnc#929883

Since commit 05a7deb1d11 (kgr: use for_each_process_thread), we have
the in_progress flag per-thread, not per-process. But we expose the
information only in /proc/<pid>/in_progress which is insufficient.
Threads might be blocking the transition and the user has no clue
which ones.

So make kgr_in_progress exposed also for threads. And make the
per-process entry return 0 only if all threads are 0.

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 fs/proc/base.c |   23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

--- a/fs/proc/base.c
+++ b/fs/proc/base.c
@@ -2136,6 +2136,26 @@ static int proc_pid_kgr_in_progress(stru
 		struct pid_namespace *ns, struct pid *pid,
 		struct task_struct *task)
 {
+	struct task_struct *t;
+	bool in_progress = false;
+
+	rcu_read_lock();
+	for_each_thread(task, t)
+		if (kgr_task_in_progress(t)) {
+			in_progress = true;
+			break;
+		}
+	rcu_read_unlock();
+
+	seq_printf(m, "%d\n", in_progress);
+
+	return 0;
+}
+
+static int proc_task_kgr_in_progress(struct seq_file *m,
+		struct pid_namespace *ns, struct pid *pid,
+		struct task_struct *task)
+{
 	seq_printf(m, "%d\n", kgr_task_in_progress(task));
 
 	return 0;
@@ -3062,6 +3082,9 @@ static const struct pid_entry tid_base_s
 	REG("projid_map", S_IRUGO|S_IWUSR, proc_projid_map_operations),
 	REG("setgroups",  S_IRUGO|S_IWUSR, proc_setgroups_operations),
 #endif
+#if IS_ENABLED(CONFIG_KGRAFT)
+	ONE("kgr_in_progress",	S_IRUSR, proc_task_kgr_in_progress),
+#endif
 };
 
 static int proc_tid_base_readdir(struct file *file, struct dir_context *ctx)
